File: 0501~1000/0981. Time Based Key-Value Store/main.py
"""
Source: https://leetcode.com/problems/time-based-key-value-store/
Date: 2023/3/18
Skill:
Runtime: 466 ms, faster than 63.29%
Memory Usage: 14.3 MB, less than 83.23%
Time complexity:
Space complexity:
Constraints:

"""

import math
from typing import List, Optional
from collections import defaultdict, Counter
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools


class TimeMap:

    def __init__(self):
        self.store = defaultdict()

    def set(self, key: str, value: str, timestamp: int) -> None:
        if key in self.store:
            val, time = self.store[key]
            if timestamp >= time:
                self.store[key] = (value, timestamp)
        else:
            self.store[key] = (value, timestamp)

    def get(self, key: str, timestamp: int) -> str:
        if key in


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)



if __name__ == "__main__":
    s = Solution()

File: 0501~1000/0981. Time Based Key-Value Store/main.cpp
/**
 * Source: tinyurl.com/y239ptxl
 * Date: 2023/10/26
 * Skill:
 * Runtime: 292 ms, faster than 55.14% of C++ online submissions
 * Memory Usage: 131.69 MB, less than 17.13% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>

using namespace std;


class TimeMap {
    unordered_map<string, map<int, string>> fmap;
public:
    TimeMap() {
    }

    void set(string key, string value, int timestamp) {
        fmap[key][timestamp] = value;
    }

    string get(string key, int timestamp) {
        auto iter = fmap[key].upper_bound(timestamp);
        if (iter == fmap[key].begin()) return "";
        return prev(iter)->second;
    }
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap* obj = new TimeMap();
 * obj->set(key,value,timestamp);
 * string param_2 = obj->get(key,timestamp);
 */


int main() {
//    Solution s;
}